### PR TITLE
Add DrawAndPrizeDistributionTimelock Smart Contract

### PR DESCRIPTION
**Pull request by kamescg**
Add the `DrawAndPrizeDistributionTimelock` to decrease the dependency on a centralized oracle for calculating and pushing PrizeDistributions off-chain.

### CODE REVIEW

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /* ============ Global Variables ============ */
+
+    /// @notice The DrawBuffer contract address.
+    IDrawBuffer public immutable drawBuffer;
+
+    /// @notice Internal PrizeDistributionFactory reference.
+    IPrizeDistributionFactory public immutable prizeDistributionFactory;
+
+    /// @notice Timelock struct reference.
+    IDrawCalculatorTimelock public timelock;
```
Comment (aodhgan) [inline contracts/DrawAndPrizeDistributionTimelock.sol:37]: can make immutable?

**Review by asselstine — APPROVED** [review]

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+  * @title  PoolTogether V4 IPrizeDistributionTimelock 
+  * @author PoolTogether Inc Team
+  * @notice The IPrizeDistributionTimelock smart contract interface...
+*/
+interface IDrawAndPrizeDistributionTimelock  {
+  function push(IDrawBeacon.Draw memory draw, uint256 totalNetworkTicketSupply) external;
+}
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:19]: Why not export these interfaces in files?

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+  * @title  PoolTogether V4 PrizeDistributionTimelock
+  * @author PoolTogether Inc Team
+  * @notice The PrizeDistributionTimelock smart contract...
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:24]: Add a description of what this contract does.

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      IDrawBeacon.Draw draw, 
+      uint256 totalNetworkTicketSupply
+    );
+
+
+  /* ============ Constructor ============ */
+
+    /**
+     * @notice Initialize DrawAndPrizeDistributionTimelock smart contract.
+     * @param _owner The smart contract owner
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:64]: Add params to Natspec doc.

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    constructor(
+      address _owner,
+      IDrawBuffer _drawBuffer,
+      IPrizeDistributionFactory _prizeDistributionFactory,
+      IDrawCalculatorTimelock _timelock) Ownable(_owner)
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:70]: Run `yarn format`

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      IDrawCalculatorTimelock _timelock) Ownable(_owner) 
+    {
+      drawBuffer = _drawBuffer;
+      prizeDistributionFactory = _prizeDistributionFactory;
+      timelock = _timelock;
+
+      emit Deployed(_drawBuffer, _prizeDistributionFactory, _timelock);
+    }
+
+  function push(IDrawBeacon.Draw memory _draw, uint256 _totalNetworkTicketSupply) external override onlyManagerOrOwner {
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:79]: Add Natspec doc.

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      timelock.lock(_draw.drawId, _draw.timestamp + _draw.beaconPeriodSeconds);
+      drawBuffer.pushDraw(_draw);
+      prizeDistributionFactory.pushPrizeDistribution(_draw.drawId, _totalNetworkTicketSupply);
+      emit DrawAndPrizeDistributionPushed(_draw.drawId, _draw, _totalNetworkTicketSupply);
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:83]: `_draw.drawId` is already logged in `_draw`. A bit redundant.

File: contracts/DrawAndPrizeDistributionTimelock.sol [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    {
+      drawBuffer = _drawBuffer;
+      prizeDistributionFactory = _prizeDistributionFactory;
+      timelock = _timelock;
+
+      emit Deployed(_drawBuffer, _prizeDistributionFactory, _timelock);
+    }
+
+  function push(IDrawBeacon.Draw memory _draw, uint256 _totalNetworkTicketSupply) external override onlyManagerOrOwner {
+      timelock.lock(_draw.drawId, _draw.timestamp + _draw.beaconPeriodSeconds);
```
Comment (PierrickGT) [inline contracts/DrawAndPrizeDistributionTimelock.sol:80]: We call `_draw.drawId` 3 times so we can store it in a variable.

File: test/DrawAndPrizeDistributionTimelock.test.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    it('should allow a push when no push has happened', async () => {
+      await drawBuffer.mock.pushDraw.returns(draw.drawId);
+      await prizeDistributionFactory.mock.pushPrizeDistribution.returns();
+      await drawCalculatorTimelock.mock.lock.returns(true);
+      await expect(drawAndPrizeDistributionTimelock.push(draw, BigNumber.from(1000000)))
+        .to.emit(drawAndPrizeDistributionTimelock, 'DrawAndPrizeDistributionPushed');
+    });
+
+    it('should not allow a push from a non-owner', async () => {
```
Comment (PierrickGT) [inline test/DrawAndPrizeDistributionTimelock.test.ts:80]: Also add a test case to be able to call the function with a manager.

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 8 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:80]: Why?

File: test/DrawAndPrizeDistributionTimelock.test.ts (hunk 9 above)
Comment (kamescg) [reply test/DrawAndPrizeDistributionTimelock.test.ts:80]: Does that test belong in this contract? Should the Owanable contract test that?

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 7 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:83]: The event is indexed by draw.drawId so, not redundant 😄

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 6 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:79]: 👍

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 5 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:70]: 👍

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 4 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:64]: 👍

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 3 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:24]: 👍

File: contracts/DrawAndPrizeDistributionTimelock.sol (hunk 2 above)
Comment (kamescg) [reply contracts/DrawAndPrizeDistributionTimelock.sol:19]: JIT coding ;) Will update interfaces once PrizeDistributorFactory is ready.

**Pull request merged**